server/pulp/plugins/rsync/configuration.py

```diff
@@ -0,0 +1,165 @@
+            elif isinstance(required_for_attr, dict):
+                for key, val in required_for_attr.iteritems():
+                    fifo.insert(0, (val, subconfig.get(key)), path + [key])
+            elif isinstance(required_for_attr, basestring):
+                if required_for_attr not in subconfig:
+                    return (False, self._err(path + [required_for_attr]))
+        return (True, None)
+
+    def _err(self, path):
+        return "%s attribute is required" % ("::".join(path))
+
+
+class RelativePathValidation(object):
+    """
+    Validates that a path does not start with a forward slash.
+    """
+    def __call__(self, value, config):
+        if os.path.isabs(value):
+            return (False, self._err(value))
+        else:
+            return (True, None)
+
+    def _err(self, value):
+        return "attribute cannot start with a /"
+
+REMOTE_MANDATORY_KEYS = {
+    "auth_type": [OneOfValidation(["publickey", "password"]),
+                  RequireOptionalIf({"remote": ["ssh_password", "ssh_user"]},
+                                    lambda x: x == "password"),
+                  RequireOptionalIf({"remote": ["ssh_identity_file", "ssh_user"]},
+                                    lambda x: x == "publickey")],
+    "host": [TypeValidation([basestring]), NonEmptyValidation()],
+    "root": [TypeValidation([basestring]), NonEmptyValidation()]
+}
+
+REMOTE_OPTIONAL_KEYS = {
+    "remote_units_path": [TypeValidation([basestring]), RelativePathValidation()],
+    "ssh_identity_file": [TypeValidation([basestring]), NonEmptyValidation()],
+    "ssh_user": [TypeValidation([basestring]), NonEmptyValidation()],
+    "ssh_password": [TypeValidation([basestring]), NonEmptyValidation()]
+}
+
+
+def validate_config(repo, config, config_conduit):
+    """
+    Validate the prospective configuration instance for the given repository.
+
+    :param repo: repository to validate the config for
+    :type  repo: pulp.plugins.model.Repository
+    :param config: configuration instance to validate
+    :type  config: pulp.plugins.config.PluginCallConfiguration
+    :param config_conduit: conduit providing access to relevant Pulp functionality
+    :type  config_conduit: pulp.plugins.conduits.repo_config.RepoConfigConduit
+    :return: tuple of (bool, str) stating that the configuration is valid or not and why
+    :rtype:  tuple of (bool, str or None)
+    """
+    _config = config.flatten()  # now config is {}
+    err_list = []
+    crit_fail = False
+    if "remote" not in _config or ("remote" in _config and not isinstance(_config["remote"], dict)):
+        err_list.append("'remote' dict missing in distributor's configuration")
+        crit_fail = True
+    if not crit_fail:
+        missing_attr_tmpl = "'%s' missing in 'remote' section of distributor's configuration"
+        for attr, validations in REMOTE_MANDATORY_KEYS.iteritems():
+            if attr not in _config["remote"]:
```

Comment:
This doesn't have to be rewritten here, but for next time consider using the [easier-to-ask-forgiveness approach](http://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html) for readability.